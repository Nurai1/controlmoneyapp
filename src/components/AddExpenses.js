import React from 'react';
import { connect } from 'react-redux';

import { addExpenses } from '../actions';
import { isPositiveNumber } from '../utilities';

const AddExpenses = ({
  generalSum,
  isCurrentSumPositive,
  dispatch,
}) => {
  let expenseName = '';
  let expenseValue = '';

  const addExpenseToList = () => {
    if (expenseName.value.trim() === '' || expenseValue.value.trim() === ''
        || !isPositiveNumber(expenseValue.value)) { return; }

    dispatch(addExpenses(Number(expenseValue.value), expenseName.value));

    expenseName.value = '';
    expenseValue.value = '';
  };

  if (isPositiveNumber(generalSum)) {
    return (
      <div className="addExpense">
        <p className="addExpense__info">
          Введите вид товара, на который вы собираетесь тратить деньги и сколько
          вы планируете на него потратить.
        </p>
        <label>
          Вид товара:
          <br />
          <input ref={(input) => { expenseName = input; }} type="text" />
          <br />
        </label>
        <label>
          Планируемые расходы:
          <br />
          <input ref={(input) => { expenseValue = input; }} type="text" />
          <br />
        </label>
        <button type="button" onClick={addExpenseToList}>Добавить расходы</button>
        <p>
          {isCurrentSumPositive ? ''
            : 'Не хватает средств, чтобы добавить категорию'}
        </p>
      </div>
    );
  }

  return '';
};

const mapStateToProps = (state) => ({
  generalSum: state.generalSum,
  isCurrentSumPositive: state.isCurrentSumPositive,
});

const AddExpensesContainer = connect(
  mapStateToProps,
)(AddExpenses);

export default AddExpensesContainer;
