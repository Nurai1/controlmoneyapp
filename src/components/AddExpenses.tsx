import React from 'react';
import { connect } from 'react-redux';

import { addExpenses } from '../store/actions';
import { isPositiveNumber } from '../utils/isPositiveNum';

import {
    AppState,
    addExpenseAction,
    ExpensePayload
} from '../store/types';

interface AddExpensesProps {
    generalSum: number,
        isCurrentSumPositive: boolean,
        addExpenses: (payload: ExpensePayload) => addExpenseAction,
}

const AddExpenses: React.FC<AddExpensesProps> = ({
  generalSum,
  isCurrentSumPositive,
                                                     addExpenses,
}) => {
  let expenseName: HTMLInputElement | null = null;
  let expenseValue: HTMLInputElement | null = null;

  const addExpenseToList = () => {
    if ((expenseName as HTMLInputElement).value.trim() === '' || (expenseValue as HTMLInputElement).value.trim() === ''
        || !isPositiveNumber((expenseValue as HTMLInputElement).value)) { return; }

    addExpenses({ value: Number((expenseValue as HTMLInputElement).value), name: (expenseName as HTMLInputElement).value });

      (expenseName as HTMLInputElement).value = '';
      (expenseValue as HTMLInputElement).value = '';
  };

  if (isPositiveNumber(generalSum)) {
    return (
      <div className="addExpense">
        <p className="addExpense__info">
          Введите вид товара, на который вы собираетесь тратить деньги и сколько
          вы планируете на него потратить.
        </p>
        <label>
          Вид товара:
          <br />
          <input ref={(input) => { expenseName = input; }} type="text" />
          <br />
        </label>
        <label>
          Планируемые расходы:
          <br />
          <input ref={(input) => { expenseValue = input; }} type="text" />
          <br />
        </label>
        <button type="button" onClick={addExpenseToList}>Добавить расходы</button>
        <p>
          {isCurrentSumPositive ? ''
            : 'Не хватает средств, чтобы добавить категорию'}
        </p>
      </div>
    );
  }

  return null;
};

const mapStateToProps = (state: AppState) => ({
  generalSum: state.generalSum,
  isCurrentSumPositive: state.isCurrentSumPositive,
});

const mapDispatchToProps = {
    addExpenses
};

const AddExpensesContainer = connect(
  mapStateToProps,
    mapDispatchToProps
)(AddExpenses);

export default AddExpensesContainer;
