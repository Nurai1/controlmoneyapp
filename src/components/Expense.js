// i think there is no need 
