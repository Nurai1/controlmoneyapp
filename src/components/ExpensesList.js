import React from 'react';
import { connect } from 'react-redux'
import { NavLink } from 'react-router-dom';

let ExpensesList = ({ expensesList }) => {

  if (expensesList!==undefined && expensesList.length !== 0){
    return(
      <ul>
        {expensesList.map((expense, idx) => {
          return (
            <li key={idx}>
              <NavLink
                exact
                to={`/${expense.id}`}>
                На категорию "{expense.name}" осталось {expense.value} рублей
              </NavLink>
            </li>
          )
        })}
      </ul>
    )
  }

  return ""
}

const mapStateToProps = (state) => {
  return {
    expensesList: state.expenses
  }
}

ExpensesList = connect(
  mapStateToProps
)(ExpensesList);

export default ExpensesList;
