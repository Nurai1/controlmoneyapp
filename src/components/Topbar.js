import React from 'react';
import { connect } from 'react-redux';

import { addGeneralSum } from '../actions';
import { isPositiveNumber } from '../utilities';

const Topbar = ({ generalSum, currentSum, dispatch }) => {
  let sumInput = '';

  const getGeneralSum = () => {
    dispatch(addGeneralSum(Number(sumInput.value)));
  };

  if (isPositiveNumber(generalSum)) {
    return (
      <div>
        <h2>Добро пожаловать в Control Money App.</h2>
        <p>
          Ваша общая сумма в наличии:
          {' '}
          {generalSum}
          {' '}
          рублей.
        </p>
        <p>
          Ваша сумма в остатке после вычета категорий:
          {' '}
          {currentSum}
          {' '}
          рублей.
        </p>
      </div>
    );
  }

  return (
    <div>
      <h2>Добро пожаловать в Control Money App.</h2>
      <h3 style={{ color: 'red' }}>
        {(generalSum === undefined || isPositiveNumber(generalSum))
          ? '' : 'Пожалуйста, введите корректное число.'}
      </h3>
      <label htmlFor="generalSum">
        Введите сумму, которой вы распологаете.
        <input ref={(input) => { sumInput = input; }} id="generalSum" type="text" />
      </label>
      <button type="button" onClick={getGeneralSum}>Начать учет</button>
    </div>
  );
};

const mapStateToProps = (state) => ({
  generalSum: state.generalSum,
  currentSum: state.currentSum,
});

const TopbarContainer = connect(
  mapStateToProps,
)(Topbar);

export default TopbarContainer;
