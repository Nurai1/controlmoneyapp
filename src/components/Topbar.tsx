import React from 'react';
import { connect } from 'react-redux';

import { addGeneralSum } from '../store/actions';
import { isPositiveNumber } from '../utils/isPositiveNum';

import {
    addGeneralSumAction,
    AppState
} from '../store/types';

interface TopbarProps {
    generalSum: number;
    currentSum: number;
    addGeneralSum: (value: number) => addGeneralSumAction;
}

export const Topbar: React.FC<TopbarProps> = ({ generalSum, currentSum, addGeneralSum }) => {
  let sumInput: HTMLInputElement | null = null;

  const getGeneralSum = () => {
      console.log('nahagahs');
      addGeneralSum(Number((sumInput as HTMLInputElement).value));
  };

  if (isPositiveNumber(generalSum)) {
    return (
      <div className="topbar">
        <h2 className="topbar__title">Добро пожаловать в Control Money App.</h2>
        <div className="topbar__counters">
          <p className="topbar__counter">
            Ваша общая сумма в наличии
            <p>(в рублях):</p>
            <span className="num-count">{generalSum}</span>
          </p>
          <p className="topbar__counter">
            Ваша сумма в остатке после вычета категорий
            <p>(в рублях):</p>
            <span className="num-count">{currentSum}</span>
          </p>
        </div>
      </div>
    );
  }

  return (
    <div className="topbar">
      <h2 className="topbar__title">Добро пожаловать в Control Money App.</h2>
      <label>
        Введите сумму, которой вы распологаете.
        <br />
        <input ref={(input) => { sumInput = input; }} type="text" />
        <br />
      </label>
      <button type="button" onClick={getGeneralSum}>Начать учет</button>
      <h3 className="topbar__mistake-text">
        {(generalSum === undefined || isPositiveNumber(generalSum))
          ? '' : 'Пожалуйста, введите корректное число.'}
      </h3>
    </div>
  );
};

const mapStateToProps = (state: AppState) => ({
  generalSum: state.generalSum,
  currentSum: state.currentSum,
});

const mapDispatchToProps = {
        addGeneralSum
    };

const TopbarContainer = connect(
  mapStateToProps,
    mapDispatchToProps,
)(Topbar);

export default TopbarContainer;
