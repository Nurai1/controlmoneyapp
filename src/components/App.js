import React from 'react';

import Topbar from './Topbar';
import AddExpenses from './AddExpenses';
import ExpensesList from './ExpensesList';

function App() {
  return (
    <div className="App wrapper">
      <Topbar />
      <AddExpenses />
      <ExpensesList />
    </div>
  );
}

export default App;
