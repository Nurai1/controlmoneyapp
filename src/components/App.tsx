import React from 'react';

import Topbar from './Topbar';
import AddExpenses from './AddExpenses';
import ExpensesList from './ExpensesList';

function App(): JSX.Element {
  return (
    <div className="App wrapper">
      <Topbar />
      <AddExpenses />
      <ExpensesList />
    </div>
  );
}

export default App;
