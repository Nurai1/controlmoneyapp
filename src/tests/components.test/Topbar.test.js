import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import pretty from "pretty";

import { Topbar } from "../../components/Topbar";

let container = null;
beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it("renders initial or full-logic page", () => {
  act(() => {
    render(<Topbar />, container);
  });

  expect(pretty(container.innerHTML)).toMatchSnapshot();

  act(() => {
    render(<Topbar generalSum={123} currentSum={123} />, container);
  });

  expect(pretty(container.innerHTML)).toMatchSnapshot();
});
