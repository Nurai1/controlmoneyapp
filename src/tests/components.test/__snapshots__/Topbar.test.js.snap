// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders initial or full-logic page 1`] = `
"<div class=\\"topbar\\">
  <h2 class=\\"topbar__title\\">Добро пожаловать в Control Money App.</h2><label>Введите сумму, которой вы распологаете.<br><input type=\\"text\\"><br></label><button type=\\"button\\">Начать учет</button>
  <h3 class=\\"topbar__mistake-text\\"></h3>
</div>"
`;

exports[`renders initial or full-logic page 2`] = `
"<div class=\\"topbar\\">
  <h2 class=\\"topbar__title\\">Добро пожаловать в Control Money App.</h2>
  <div class=\\"topbar__counters\\">
    <p class=\\"topbar__counter\\">Ваша общая сумма в наличии<p>(в рублях):</p><span class=\\"num-count\\">123</span></p>
    <p class=\\"topbar__counter\\">Ваша сумма в остатке после вычета категорий<p>(в рублях):</p><span class=\\"num-count\\">123</span></p>
  </div>
</div>"
`;
